File: cpp/0077组合.cpp
#include<iostream>
#include<vector>

using namespace std;

vector<vector<int>> res;
vector<int> temp;


void back_tracking(int s,int n,int k)
{
    if(k<1)
    {
        // 值传递
        res.push_back(temp);
        return;
    }

    for (;s<=n-k+1;++s)
    {
        temp.push_back(s);
        back_tracking(s + 1, n, k-1);
        temp.pop_back();
    }
}
vector<vector<int>> combine(int n, int k)
{
    back_tracking(1, n, k);

    return res;
}

File: python/main/0077组合.py



def combine(n: int, k: int) -> list:
    res=[]
    
    def _combine(s:int,n:int,k:int,temp:list):
        if(k<1):
            # print(temp)
            res.append(temp.copy())
            return

        for i in range(s,n-k+2):
            temp.append(i)
            _combine(i+1,n,k-1,temp)
            temp.pop()

    _combine(1,n,k,[])
    return res


print(combine(5,3))
